### PR TITLE
Add JSON path example

### PR DESCRIPTION
**Pull request by akupila**
As i was [stumbling with the basics](https://github.com/alecthomas/participle/issues/72), hopefully a super simple example can help others get started. 

This PR adds a new example for parsing a json path and applying that to a json file, a bit like an interactive but primitive version of [jq](https://stedolan.github.io/jq/).

Demo: https://asciinema.org/a/nSWGX6ZPghlpG8blXwGvYmui5

### CODE REVIEW

File: _examples/json/main.go [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+	var input map[string]interface{}
+	err = json.NewDecoder(r).Decode(&input)
+	ctx.FatalIfErrorf(err)
+
+	enc := json.NewEncoder(os.Stdout)
+	enc.SetIndent("", "  ")
+
+	for {
+		// Example: check_run.check_suite.pull_requests[0].head.sha
+		fmt.Fprint(os.Stderr, "Enter a json path: ")
```
Comment (alecthomas) [inline _examples/json/main.go]: I'd just make this an argument ge. `json <query> <files...>`

File: _examples/json/main.go [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+		q, err := bufio.NewReader(os.Stdin).ReadString('\n')
+		ctx.FatalIfErrorf(err)
+
+		var expr pathExpr
+		if err := parser.ParseString(q, &expr); err != nil {
+			fmt.Fprintln(os.Stderr, err)
```
Comment (alecthomas) [inline _examples/json/main.go]: Use `ctx.FatalIfErrorf(err)` here.

File: _examples/json/main.go [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+			continue
+		}
+
+		result, err := match(input, expr)
+		if err != nil {
```
Comment (alecthomas) [inline _examples/json/main.go]: Same here.

File: _examples/json/main.go [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+			return nil, fmt.Errorf("%q is not a map", e.Obj)
+		}
+		v = v.MapIndex(reflect.ValueOf(e.Obj))
+		if !v.IsValid() {
+			return nil, fmt.Errorf("not found: %q", e.Obj)
+		}
+		if v.Kind() == reflect.Interface {
+			v = v.Elem()
+		}
+		for _, a := range e.Acc {
```
Comment (alecthomas) [inline _examples/json/main.go]: Nice!

File: _examples/json/main.go [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+		if err != nil {
+			fmt.Fprintln(os.Stderr, err)
+			continue
+		}
+		enc.Encode(result)
+	}
+}
+
+func match(input map[string]interface{}, expr pathExpr) (interface{}, error) {
+	v := reflect.ValueOf(input)
```
Comment (alecthomas) [inline _examples/json/main.go]: I'd probably just use type switches here, but up to you.

File: _examples/json/main.go (hunk 1 above)
Comment (akupila) [reply _examples/json/main.go]: I made it so it's not interactive any more. Instead you pass in a query and 1+ files

File: _examples/json/main.go (hunk 2 above)
Comment (akupila) [reply _examples/json/main.go]: The idea previously was that the app should not exit on error, but instead ask for a new query. Now however as it's no longer interactive, i removed the kong dependency to keep thing simpler and just do `os.Exit(1)` instead.

File: _examples/json/main.go (hunk 5 above)
Comment (akupila) [reply _examples/json/main.go]: Done. Not sure if i find the non-reflect code easier to read but maybe it doesn't stand out as much to however is reading the example.

**Comment by akupila** [comment]
I addressed your comments with a couple changes to how this works:
- It is no longer interactive. Instead you pass in the the query and files to query and it'll execute it (more like `jq`). For example `go run . "check_run.check_suite.pull_requests[0].head.sha" github-webhook.json`
- As it is no longer interactive, any error will immediately terminate
- The matching is done with type switches rather than using the reflect package.
- The output is encoded as json only if the match is for an object. Not important for the example but i find this a bit nicer to use

File: _examples/json/main.go [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+	Parts []part `@@ { "." @@ }`
+}
+
+type part struct {
+	Obj string `@Ident`
+	Acc []acc  `("[" @@ "]")*`
+}
+
+type acc struct {
+	Name  *string `@(String|Char|RawString)`
```
Comment (akupila) [inline _examples/json/main.go]: Is this correct? Would `@String` be sufficient?

Comment (alecthomas) [reply _examples/json/main.go]: Yep `@String` should be sufficient.

**Comment by alecthomas** [comment]
Thanks!

**Comment by itchyny** [comment]
How about renaming the directory to `json_path` so that the visitor does not confuse the example parses JSON? And maybe adding an example for JSON parser?

**Pull request merged**